core/src/main/java/com/linecorp/armeria/common/stream/PathStreamMessage.java

```diff
@@ -316,8 +326,15 @@ public void completed(Integer result, ByteBuf byteBuf) {
                             byteBuf.release();
                         }
                         downstream.onNext(data);
-                        reading = false;
-                        read();
+                        final long position = this.position;
+                        assert position <= end;
+                        if (position < end) {
+                            reading = false;
+                            read();
+                        } else {
+                            maybeCloseFileChannel();
```

Comment:
Note: the stream will close if the end of the file is reached regardless of `PathStreamMessage#end`